[CSS] New property page for animation-composition

dipikabh opened the pull request:
<!-- 👀 Thanks for opening a PR! Read comments like this one to get your PR merged faster. -->
#### Summary
<!-- ✍️ In a sentence or two, describe your changes -->
- This PR adds documentation for the new CSS property `animation-composition`, supported in FF104.
- This property is behind a preference in FF104, so you would need to enable `layout.css.animation-composition.enabled` in Beta or Nightly to see the 'Example' in action.

#### Notes

- This property is not yet supported in the shorthand `animation` property.
- The property is not yet implemented in Chrome and not shipped in Safari.
- Because of its limited support right now, I am going to put in backlog the following tasks:
   - adding an interactive example
   - updating other `animation` property pages with link to `animation-composition`

#### Supporting details
<!-- 🔗 Link to supporting information, such as bug trackers, source control, release notes, and vendor docs. -->
Bug tracker: https://bugzilla.mozilla.org/show_bug.cgi?id=1293490
Doc tracker: https://github.com/mdn/content/issues/18771
Spec: https://drafts.csswg.org/css-animations-2/#animation-composition
PR for BCD update: https://github.com/mdn/browser-compat-data/pull/17518
PR for experimental feature update: https://github.com/mdn/content/pull/19849

#### Metadata
<!-- ✅ Check a box if applicable, like this: [x]

This PR…
-->
- [x] Adds a new document
- [ ] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

<!-- 👷‍♀️ After submitting, review the results of the "Checks" tab! -->

dipikabh committed:
Adds the file for animation-composition

dipikabh committed:
Adds documentation for animation-composition

dipikabh committed:
Makes minor revisions

estelle reviewed (commented):
i have to read the spec, so take these comments with a grain of salt. Will re-review tomorrow.

estelle commented on files/en-us/web/css/animation-composition/index.md:
````diff
@@ -0,0 +1,166 @@
+```css
+.icon:hover {
+  filter: blur(5px);
+  animation: brightness-pulse 3s infinite;
````
i found this error in the specification, so PR'ed a fix. That said, I think our example should be different from the spec so users can see a few valid use cases rather than one (currently invalid) repeated

dipikabh replied:
If we do specify the two animations like so:

>   animation: brightness 3s infinite alternate, pulse 2s infinite alternate;

Then I'd also have to add another `keyframes` at-rule for brightness, correct? There's only one right now for pulse. 

The other alternative is to update the 'animation' property to match `@keyframes pulse`:

>   animation: pulse 3s infinite alternate;

estelle replied:
The `animation-name` did not match the `@keyframes <name>`. The merged the PR. I thought you were doing two animations, 1: brightness, and 2: pulse. But you and the spec authors were just doing one.

dipikabh replied:
I just updated the animation shorthand to place the keyframe name at the end: `animation: 3s infinite alternate pulse;`

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+  height: 50px;
+  background: green;
+  border-radius: 10px;
+  transform: translateX(50px) rotate(90deg);
```
```suggestion
  transform:  translateX(50px) rotate(45deg) ;
```

It is a little more obvious what is happening to translateX at 45deg than 90deg, because 90 is vertical.

estelle commented on files/en-us/web/css/animation-composition/index.md:
````diff
@@ -0,0 +1,166 @@
+
+```css
+@keyframes slide {
+  from {transform: translateX(100px);}
````
```suggestion
  20%, 40% {transform: translateX(100px);}
```

dipikabh replied:
I've made the suggested change but I feel it is adding more complexity to the use case. There will be too many things to unpack.

estelle commented on files/en-us/web/css/animation-composition/index.md:
````diff
@@ -0,0 +1,166 @@
+```css
+@keyframes slide {
+  from {transform: translateX(100px);}
+  to {transform: translateX(150px);}
````
```suggestion
  80%, 100% {transform: translateX(150px);}
```

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+  background: green;
+  border-radius: 10px;
+  transform: translateX(50px) rotate(90deg);
+  animation: slide 2s linear infinite;
```
```suggestion
  animation: slide 5s linear infinite;
```

Provide a little more time so people can grok the effect. Also, I used 20/40/80/100% so stuff lasts 1s

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+
+{{EmbedLiveSample("Reversing the animation direction","100%","250")}}
+
+- With `replace`, the composite effect value in the `from` keyframe is `transform: translateX(100px)` (replacing `translateX(50px) rotate(90deg)`).
```
```suggestion
- With `replace`, the composite effect value in the `20%` keyframe is `transform: translateX(100px)` (completely replacing `translateX(50px) rotate(45deg)`).
```

dipikabh replied:
This is great, thanks!

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+{{EmbedLiveSample("Reversing the animation direction","100%","250")}}
+
+- With `replace`, the composite effect value in the `from` keyframe is `transform: translateX(100px)` (replacing `translateX(50px) rotate(90deg)`).
+- With `add`, the composite effect value in the `from` keyframe is `translateX(50px) rotate(90deg)` followed by `transform: translateX(100px)`.
```
```suggestion
- With `add`, the composite effect value in the `20%` keyframe is `translateX(50px) rotate(45deg)` followed by `transform: translateX(100px)`.
```

this still doesn't make sense to me. You can't have two translate values at a time. That doesn't mean it's not possible, I am just not grokking it.

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+
+- With `replace`, the composite effect value in the `from` keyframe is `transform: translateX(100px)` (replacing `translateX(50px) rotate(90deg)`).
+- With `add`, the composite effect value in the `from` keyframe is `translateX(50px) rotate(90deg)` followed by `transform: translateX(100px)`.
+- With `accumulate`, the composite effect value in the `from` keyframe is `translateX(150px) rotate(90deg)`.
```
```suggestion
- With `accumulate`, the composite effect value in the `20%` keyframe is `translateX(150px) rotate(45deg)`.
```

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+Consider different values for the `animation-composition` property in the above example. The final effect value in each of those cases will be calculated as explained below:
+
+- With `replace`, `blur(10px)` will replace `blur(5px)` in the `0%` keyframe. This is the default behavior of the property.
+- With `add`, the composite effect value in `0%` keyframe will be `blur(5px) blur(10px)`.
```
```suggestion
- With `add`, the composite effect value in the `0%` keyframe will be `blur(5px) blur(10px)`.
```

estelle replied:
```suggestion
- With `add`, the composite effect value in the `0%` keyframe will be `blur(5px) blur(10px)`.
```

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+### Values
+
+- `replace`
+  - : The effect value overrides the underlying value of the property. This is the default composite operation.
```
```suggestion
  - : The animated value overrides any non-`!important` underlying values of the property. This is the default value.
```

dipikabh replied:
We do need to use the terms "effect value" and "underlying value"...

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+- `replace`
+  - : The effect value overrides the underlying value of the property. This is the default composite operation.
+- `add`
+  - : The effect value builds on the underlying value of the property. This values produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.
```
```suggestion
  - : The animated value builds on the underlying value of the property. This value produces an additive effect. For property types where it is not possible to add multiple values, the animated value overrides the non-!important non-animated value. 
```

dipikabh replied:
I am going to retain the terms "effect value" and "underlying value". I have tried to explain them in the "Description" section.
I'll need to cross-check if the edits to the sentence "For property types..." convey the intended meaning here.

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+  border-radius: 10px;
+  transform: translateX(50px) rotate(90deg);
+  animation: slide 2s linear infinite;
+}
```
```suggestion
}
.target:hover {
   animation-play-state: paused;
 }
```

estelle replied:
Enable pausing on hover so users can see the effect, especially of to color.

estelle replied:
before adding the color explanation, let's see if the other edits make sense. The color additives are cool. we should likely explain how the r, g, and b values are added or replaced if that is in fact what is happening. i have gone down this rabbit hole too long today...

dipikabh commented:
Thanks for your review, @estelle! I'll start fixing your comments now.

dipikabh commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+  - CSS
+  - CSS Animations
+  - CSS Property
+  - Reference
```
Yeah, I would leave this because we'll be getting rid of tags

dipikabh commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+---
+{{CSSRef}}
+
+The **`animation-composition`** [CSS](/en-US/docs/Web/CSS) property specifies the composite operation to use when multiple animations affect the same property simultaneously.
```
Thanks for pointing this - I'm adding a glossary page for 'composite operation'.

The spec itself redirects to https://drafts.csswg.org/web-animations-1/#composite-operation for the term. This term in glossary will also be useful for [KeyframeEffect.composite](https://developer.mozilla.org/en-US/docs/Web/API/KeyframeEffect/composite).

estelle replied:
Excellent. thanks.

dipikabh commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+- `replace`
+  - : The effect value overrides the underlying value of the property. This is the default composite operation.
+- `add`
+  - : The effect value builds on the underlying value of the property. This values produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.
```
I do not know the answer to this.

Not sure if this will help or is related, but I had found this earlier:

> "The most important thing to be aware of is that properties that can be typically accelerated, such as "opacity", "filter", "transform", "translate", "scale" and "rotate" will not be fully accelerated if composite operations are used. Instead, the animation resolution will operate on the main thread, while the application of the property will operate on a GraphicsLayer. We are continuing to look into how to address this in the future.

(source: https://bugs.webkit.org/show_bug.cgi?id=189299)

dipikabh commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,166 @@
+- `add`
+  - : The effect value builds on the underlying value of the property. This values produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.
+- `accumulate`
+  - : The effect and underlying values are combined. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.
```
- I used the word 'combined' here and not 'added' to distinguish from the other operation named `add`. So to describe `add`, I used 'builds on' and for `accumulate`, I used 'combined', just a synonym that is not the word add.
- The word 'commutative' has not been defined or linked in the spec and I am not sure if they really mean a simple add, so I retained commutative.
- From what I understand, the order between the `underlying value` and `effect value` is important in a case like this (was shared by the engineer):

```css
@keyframes anim {
  from { transform: translateX(50px); }
  to { ... }
}
#target {
  transform: translateX(100px) rotate(90deg);
}
```

Assume that the `animation-composition` specified in `#target` is `add`.
The underlying value for `transform` here is `translateX(100px) rotate(90deg)` and the effect value is `translateX(50px)`

Since `add` combines underlying value and then effect value, so the effective "from" value will be `transform: translateX(100px) rotate(90deg) translateX(50px)` and not `transform: translateX(50px) translateX(100px) rotate(90deg)`.
It is not really the order of animations but how the final effect value is calculated.

Let me know if you were trying to make a different point.

estelle replied:
your original text is fine. I am being too detailed. We can add all the detail when it's actually supported and we can see what is going on in the debugger (right now, dev tools aren't showing animation tools when composition is set)

dipikabh committed:
Fixes review comments and adds a glossary page

dipikabh commented:
Hi @estelle, I have incorporated most of your suggestions. For some others, I've left a comment or a question.

I've updated the example though I fear it might be too many behaviors to unpack at the same time while also trying to understand a new property. We could probably retain the original 'simple' example and add another one with your suggested enhancements. What do you think?

estelle commented on files/en-us/glossary/composite_operation/index.md:
```diff
@@ -0,0 +1,19 @@
+- **replace**: The effect value replaces the underlying value. The final effect value in this case is the original effect value itself.
+- **add**: The effect value is added to the underlying value.
+- **accumulate**: The effect value is combined with the underlying value.
+
```
```suggestion

> **Note:** Currently, compostion in CSS only applies to animations. There has been discussion about more general additive syntax, but as of late 2022, there are no other CSS features to which `add` and `accumulate` apply.
```

estelle replied:
not sure if you want to add a note...

dipikabh replied:
I've added it for now. Let me check on the guidance for adding such time-specific info - if needed, I'll remove it

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,175 @@
+---
+{{CSSRef}}
+
+The **`animation-composition`** [CSS](/en-US/docs/Web/CSS) property specifies the composite operation to use when multiple animations affect the same property simultaneously.
```
```suggestion
The **`animation-composition`** CSS property specifies the [composite operation](/en-US/docs/glossary/composite_operation) to use when multiple animations affect the same property simultaneously.
```

estelle commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,175 @@
+- `replace`
+  - : The effect value overrides the underlying value of the property. This is the default value.
+- `add`
+  - : The effect value builds on the underlying value of the property. This operation produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.
```
```suggestion
  - : The effect value builds on the underlying value of the property. This operation produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is the underlying value followed by the effect value.
```

estelle commented on files/en-us/web/css/animation-composition/index.md:
````diff
@@ -0,0 +1,175 @@
+```css
+.icon:hover {
+  filter: blur(5px);
+  animation: brightness-pulse 3s infinite;
````
```suggestion
  animation: pulse 3s infinite;
```

estelle committed:
Update files/en-us/web/css/animation-composition/index.md

estelle commented on files/en-us/web/css/animation-composition/index.md:
````diff
@@ -0,0 +1,175 @@
+
+```css
+@keyframes slide {
+  20%, 40% {
````
```suggestion
  0%, 20% {
```

estelle replied:
i am reverting the change i made a few days ago. We need to have it at 0% because the content is talking about what happens at the 0% mark. we should still hold it there of the 20% so people can see the difference between the 3.

dipikabh replied:
The explanation text for this example does talk about the '20%, 40%' keyframe. This is in the 'Examples' section.
The one with the '0%' keyframe is a different example in the 'Description' section. 
So I'm keeping this as is `20%, 40%`.

estelle replied:
the reason to change it back to 0% is so it is additive / cumulative from the start. If we start at 20%, then the 0% is the original value, and the effect is less visible. I made a mistake when I suggested changing it to 20%

dipikabh replied:
fixing this here: https://github.com/mdn/content/pull/20250

estelle reviewed (commented):
Looking good. just a few edits.

dipikabh committed:
Fixes review comments

dipikabh committed:
Updates note to remove time-specific info

bsmth commented on files/en-us/web/css/animation-composition/index.md:
```diff
@@ -0,0 +1,175 @@
+<!-- {{EmbedInteractiveExample("pages/css/animation-composition.html")}} -->
+
+<!--only longhand animation-composition is enabled in FF104, so commenting out the sentence below for now. Not removing the sentence because it can be uncommented when support in shorthand animation property is added.-->
+<!-- It is often convenient to use the shorthand property {{cssxref("animation")}} to set all animation properties at once.-->
```
I think these comments can be removed and we can create an issue or a TODO somewhere to track this

dipikabh replied:
Yes! Thanks for pointing these out. I've removed the comments. For future tracking, I've added it to our 'to triage' list.

dipikabh committed:
Removes todo comments

bsmth reviewed (approved):
Thank you! 🙌🏻
